subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/kubernetes/KubernetesReservationClient.kt

```diff
@@ -65,46 +64,44 @@ class KubernetesReservationClient(
     override suspend fun claim(
         reservations: Collection<Reservation.Data>,
         serialsChannel: SendChannel<Serial>,
-        reservationDeployments: SendChannel<String>
+        reservationDeployments: SendChannel<String> // TODO: make this state internal
     ) {
         if (state !is State.Idling) {
             val error = RuntimeException("Unable to start reservation job. Already started")
             logger.critical(error.message.orEmpty())
             throw error
         }
-
         logger.debug("Starting deployments for configuration: $configurationName...")
-
-        val podsChannel: Channel<Pod> = reservations
-            .iterateInParallel { _, reservation ->
-                val deploymentName = generateDeploymentName()
-                reservationDeployments.send(deploymentName)
-
-                logger.debug("Starting deployment: $deploymentName")
-                when (reservation.device) {
-                    is Device.Emulator -> getEmulatorDeployment(
-                        emulator = reservation.device,
-                        deploymentName = deploymentName,
-                        count = reservation.count
-                    )
-                    is Device.Phone -> getDeviceDeployment(
-                        count = reservation.count,
-                        phone = reservation.device,
-                        deploymentName = deploymentName
-                    )
-                    is Device.LocalEmulator -> throw IllegalStateException(
-                        "Local emulator ${reservation.device} is unsupported in kubernetes reservation"
-                    )
-                }.create()
-                logger.debug("Deployment created: $deploymentName")
-
-                listenPodsFromDeployment(
+        val podsChannel = Channel<Pod>()
+        state = State.Reserving(pods = podsChannel)
+
+        reservations.forEach { reservation ->
+            val deploymentName = generateDeploymentName()
+            reservationDeployments.send(deploymentName)
+
+            logger.debug("Starting deployment: $deploymentName")
+            when (reservation.device) {
+                is Device.Emulator -> getEmulatorDeployment(
+                    emulator = reservation.device,
+                    deploymentName = deploymentName,
+                    count = reservation.count
+                )
+                is Device.Phone -> getDeviceDeployment(
+                    count = reservation.count,
+                    phone = reservation.device,
                     deploymentName = deploymentName
                 )
-            }
-            .merge()
```

Comment:
`merge` function did all work in unconfined coroutines (globalscope == default dispatcher)

```kotlin
suspend fun <T> Collection<Channel<T>>.merge(): Channel<T> = Channel<T>().apply {
    forEach { channel ->
        GlobalScope.launch {
            for (item in channel) {
                send(item)
            }
        }
    }
}
```
We've closed resulting channel in the end, but cancellation was not propagated properly to the original coroutines.
Therefore, podsChannel was always open.

```kotlin
listenPodsFromDeployment(
  GlobalScope.launch {
    ....
    while (!podsChannel.isClosedForSend && pods.isNotEmpty()) {
```
It worked only due to empty pods. After deleting deployment the `podsFromDeployment` fails and returns empty list. It breaks the loop.

This is why I've used only one channel to communicate.